core/ast_node.js

```diff
@@ -0,0 +1,657 @@
+  if (curIdx <= -1) {return;}
+  var nxtIdx = curIdx + 1;
+  var newAstNode = null;
+
+  for (var i = nxtIdx; i < inputs.length; i++) {
+    var newInput = inputs[i];
+    var field = this.findNextEditableField_(location, newInput, true);
+    if (field) {
+      newAstNode = Blockly.ASTNode.createFieldNode(field);
+      break;
+    } else if (newInput.connection) {
+      var connection = newInput.connection;
+      newAstNode = Blockly.ASTNode.createConnectionNode(connection);
+      break;
+    }
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the next input or field given a field location.
+ * @param {!Blockly.Field} location Current location in the ast.
+ * @param {!Blockly.Input} parentInput The parent input of the field.
+ * @return {?Blockly.ASTNode} The ast node holding the next field or connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findNextForField_ = function(location, parentInput) {
+  var newAstNode = this.findNextEditableField_(location, parentInput);
+
+  if (!newAstNode || !newAstNode.getLocation()) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(parentInput.connection);
+  }
+  return newAstNode;
+};
+
+/**
+ * Given the current selected field or connection find the previous connection
+ *     or field.
+ * @param {!Blockly.Connection} location Current location in the ast.
+ * @param {!Blockly.Input} parentInput Parent input of the connection or field.
+ * @return {?Blockly.ASTNode} The ast node holding the previous field or
+ *     connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findPrevForInput_ = function(location, parentInput){
+  var block = location.sourceBlock_;
+  var inputs = block.inputList;
+  var curIdx = inputs.indexOf(parentInput);
+  var newAstNode = null;
+
+  for (var i = curIdx; i >= 0; i--) {
+    var newInput = inputs[i];
+    var fieldNode = this.findPreviousEditableField_(location, newInput, true);
+    if (newInput.connection && newInput.connection !== parentInput.connection) {
+      newAstNode = Blockly.ASTNode.createInputNode(newInput);
+      break;
+    } else if (fieldNode && fieldNode.getLocation() && fieldNode !== location) {
+      newAstNode = Blockly.ASTNode.createFieldNode(fieldNode.getLocation());
+      break;
+    }
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the previous input or field given a field location.
+ * @param {!Blockly.Field} location The current location of the cursor.
+ * @param {!Blockly.Input} parentInput The parent input of the field.
+ * @return {?Blockly.ASTNode} The ast node holding the previous input or field.
+ * @private
+ */
+Blockly.ASTNode.prototype.findPrevForField_ = function(location, parentInput) {
+  var block = location.sourceBlock_;
+  var inputs = block.inputList;
+  var curIdx = inputs.indexOf(parentInput);
+  var prevLocation = this.findPreviousEditableField_(location, parentInput);
+  var newAstNode;
+
+  if (!prevLocation && curIdx - 1 >= 0) {
+    newAstNode = Blockly.ASTNode.createInputNode(inputs[curIdx - 1]);
```

Comment:
Does this need to continue across the tree (by checking for a previous input, then a field if the previous input is a dummy, and so on within the block) or do you handle that at a higher level?

It seems like you'd get stuck at the first field of an input using this code.